File: layers/perl/ad_layer_line_delete_in_place.pl
#!/usr/bin/env perl
# ad_layer_line_delete_in_place.pl — Delete content BEFORE joining lines.
#
# Perl twin of ad_layer_line_delete_in_place.c.
# Two modes: --mode batch (default) or --mode interleaved.
#
# Sliding-window algorithm (Phase 2 fix, handles arbitrary N lines):
#   Walk the ops. When a join_lines(L) is followed by content deletes
#   (first delete at col 1), start a block. Within the block, match
#   join + delete+ repeatedly. Each match:
#     - Emits content deletes at L + 1 + offset.
#     - Defers the join to a pending list.
#     - Advances offset by 1.
#   When the block ends, emit all pending joins.
#
# See C source for full documentation.

use strict;
use warnings;
use utf8;

binmode(STDIN,  ':utf8');
binmode(STDOUT, ':utf8');
binmode(STDERR, ':utf8');

my $ldi_mode = 0;  # 0=batch, 1=interleaved

# Parse args
for (my $ai = 0; $ai < @ARGV; $ai++) {
    if ($ARGV[$ai] eq '--mode' && $ai + 1 < @ARGV) {
        $ai++;
        $ldi_mode = 1 if $ARGV[$ai] eq 'interleaved';
    } elsif ($ARGV[$ai] =~ /^--mode=(.*)/) {
        $ldi_mode = 1 if $1 eq 'interleaved';
    } elsif ($ARGV[$ai] eq '--help' || $ARGV[$ai] eq '-h') {
        print STDERR "ad_layer_line_delete_in_place — delete content before joining lines\n\n";
        print STDERR "Usage: ad_layer_line_delete_in_place [--mode batch|interleaved] < ops.tsv\n\n";
        print STDERR "Options:\n";
        print STDERR "  --mode batch        Delete all content first, then join all (default)\n";
        print STDERR "  --mode interleaved  Delete each line then immediately join\n";
        print STDERR "  --help, -h          Show this help\n";
        exit 0;
    }
}

# --- TSV parse / write ---------------------------------------------------

sub parse_op {
    my ($line) = @_;
    chomp $line;
    my @f = split /\t/, $line, -1;

    # Standard 4-field format: type\tline\tcol\tcode (optionally \tchar_repr)
    if (@f >= 4
        && $f[0] ne 'insert_line' && $f[0] ne 'batch_insert'
        && $f[0] ne 'delete_line' && $f[0] ne 'keep_line'
        && $f[0] ne 'join_lines'  && $f[0] ne 'split_line') {
        return {
            type => $f[0],
            line => $f[1] + 0,
            col  => $f[2] + 0,
            code => $f[3] + 0,
        };
    }

    # insert_line format: insert_line\t<line>\t<text>
    if (@f >= 3 && $f[0] eq 'insert_line') {
        return {
            type => 'insert_line',
            line => $f[1] + 0,
            col  => 0,
            code => 0,
            text => $f[2],
        };
    }

    # batch_insert format: batch_insert\t<line>\t<col>\t<codes>
    if (@f >= 4 && $f[0] eq 'batch_insert') {
        return {
            type => 'batch_insert',
            line => $f[1] + 0,
            col  => $f[2] + 0,
            code => 0,
            text => $f[3],
        };
    }

    # split_line format: split_line\t<line>\t<col>
    if (@f >= 3 && $f[0] eq 'split_line') {
        return {
            type => 'split_line',
            line => $f[1] + 0,
            col  => $f[2] + 0,
            code => 0,
        };
    }

    # keep_line / join_lines / delete_line format: <type>\t<line>
    if (@f >= 2
        && ($f[0] eq 'keep_line' || $f[0] eq 'join_lines'
            || $f[0] eq 'delete_line')) {
        return {
            type => $f[0],
            line => $f[1] + 0,
            col  => 0,
            code => 0,
        };
    }

    # Catch-all: store the raw line for verbatim pass-through.
    if (@f >= 1) {
        return {
            type => $f[0],
            line => $f[1] ? ($f[1] + 0) : 0,
            col  => $f[2] ? ($f[2] + 0) : 0,
            code => $f[3] ? ($f[3] + 0) : 0,
            text => $line,
            raw  => 1,
        };
    }

    return undef;
}

sub char_repr {
    my ($code) = @_;
    return "\\n"     if $code == 10;
    return "\\t"      if $code == 9;
    return "\\r"       if $code == 13;
    return "space"     if $code == 32;
    return "'" . chr($code) . "'" if $code >= 33 && $code <= 126;
    return "$code";
}

sub write_op {
    my ($op) = @_;
    if ($op->{type} eq 'keep_line') {
        printf "keep_line\t%d\n", $op->{line};
    } elsif ($op->{type} eq 'join_lines') {
        printf "join_lines\t%d\n", $op->{line};
    } elsif ($op->{type} eq 'split_line') {
        printf "split_line\t%d\t%d\n", $op->{line}, $op->{col};
    } elsif ($op->{type} eq 'insert_line') {
        printf "insert_line\t%d\t%s\n", $op->{line}, $op->{text} // '';
    } elsif ($op->{type} eq 'delete_line') {
        printf "delete_line\t%d\n", $op->{line};
    } elsif ($op->{type} eq 'batch_insert') {
        printf "batch_insert\t%d\t%d\t%s\n",
            $op->{line}, $op->{col}, $op->{text} // '';
    } elsif ($op->{raw}) {
        printf "%s\n", $op->{text};
    } else {
        printf "%s\t%d\t%d\t%d\t%s\n",
            $op->{type}, $op->{line}, $op->{col}, $op->{code},
            char_repr($op->{code});
    }
}

sub is_line_op {
    my ($op) = @_;
    return defined $op && (
        $op->{type} eq 'keep_line' || $op->{type} eq 'join_lines' ||
        $op->{type} eq 'split_line' || $op->{type} eq 'delete_line' ||
        $op->{type} eq 'batch_insert'
    );
}

# --- Layer transform (batch mode) ---------------------------------------

sub transform_hunk {
    my ($ops) = @_;
    my @in = @$ops;
    my $n = scalar @in;
    my @out;

    my $i = 0;
    while ($i < $n) {
        # Check: is in[i] a join_lines followed by content deletes at col 1?
        if ($in[$i]{type} eq 'join_lines'
            && $i + 1 < $n
            && $in[$i+1]{type} eq 'delete'
            && !is_line_op($in[$i+1])
            && $in[$i+1]{col} == 1) {

            # ── Sliding-window block ──
            my $line_off = 0;
            my @pending_joins;
            my $j = $i;

            while ($j < $n && $in[$j]{type} eq 'join_lines') {
                my $join_line = $in[$j]{line};

                # Scan content deletes. First must be at col 1.
                my $ce = $j + 1;
                if ($ce < $n
                    && $in[$ce]{type} eq 'delete'
                    && !is_line_op($in[$ce])
                    && $in[$ce]{col} == 1) {
                    $ce++;
                    while ($ce < $n
                           && $in[$ce]{type} eq 'delete'
                           && !is_line_op($in[$ce])) {
                        $ce++;
                    }
                }

                if ($ce > $j + 1) {
                    # Content found (first at col 1). Emit at L+1+line_off.
                    for my $k ($j + 1 .. $ce - 1) {
                        my %op = %{$in[$k]};
                        $op{line} = $join_line + 1 + $line_off;
                        push @out, \%op;
                    }
                    push @pending_joins, $in[$j];
                    $line_off++;
                    $j = $ce;
                } else {
                    # No content — trailing join. End block.
                    push @pending_joins, $in[$j];
                    $j++;
                    last;
                }
            }

            # Emit all pending joins.
            push @out, @pending_joins;
            $i = $j;
            next;
        }

        # No match — emit unchanged.
        push @out, $in[$i];
        $i++;
    }

    return \@out;
}

# --- Main: hunk-by-hunk driver -------------------------------------------

my $in_hunk = 0;
my $hunk_count = 0;
my @hunk_ops;
my ($hunk_target, $hunk_del, $hunk_ins, $hunk_end_ins, $hunk_end_del) =
    (0, 0, 0, 0, 0);

sub flush_hunk {
    return unless @hunk_ops;

    my $out_ops;
    if ($ldi_mode == 1) {
        # Interleaved mode: pass through.
        $out_ops = \@hunk_ops;
    } else {
        $out_ops = transform_hunk(\@hunk_ops);
    }

    printf "HUNK\t%d\t%d\t%d\t%d\t%d\n",
        $hunk_target, $hunk_del, $hunk_ins, $hunk_end_ins, $hunk_end_del;
    for my $op (@$out_ops) {
        write_op($op);
    }
    print "HUNK_END\n";
    @hunk_ops = ();
}

while (my $line = <STDIN>) {
    chomp $line;
    next if $line eq '';

    if ($line =~ /^#/) {
        if ($line =~ /raw diff|post-processed/) {
            print "# diffvim post-processed v2\n";
        } else {
            print "$line\n";
        }
        next;
    }

    if ($line =~ /^HUNK\t(\d+)\t(\d+)\t(\d+)\t(\d+)\t(\d+)/) {
        flush_hunk() if $in_hunk;
        ($hunk_target, $hunk_del, $hunk_ins, $hunk_end_ins, $hunk_end_del) =
            ($1, $2, $3, $4, $5);
        $in_hunk = 1;
        $hunk_count++;
        next;
    }

    if ($line =~ /^HUNK_END/) {
        flush_hunk() if $in_hunk;
        $in_hunk = 0;
        next;
    }

    last if $line eq 'EOF';

    if ($in_hunk) {
        my $op = parse_op($line);
        push @hunk_ops, $op if $op;
    }
}

flush_hunk() if $in_hunk && @hunk_ops;

print "\n";
exit 0;

File: layers/c/ad_layer_line_delete_in_place.c
/* ad_layer_line_delete_in_place.c — Delete content BEFORE joining lines.
 *
 * When the diff engine deletes multiple consecutive lines, it produces:
 *
 *   delete(line1 chars) -> join_lines -> delete(line2 chars) -> join_lines -> ...
 *
 * The join_lines pulls line2's content UP to line1 before it's deleted.
 * Visually, the user sees content jumping up before disappearing.
 *
 * Two algorithms are available, switchable via --mode:
 *
 * --mode batch (default):
 *   Delete ALL content first, THEN join all empty lines.
 *   delete(line1) -> delete(line2 at L+1) -> delete(line3 at L+2) ->
 *   ... -> join -> join -> ...
 *   The content deletes for line N are emitted at line L+N-1 (their
 *   original pre-join position). All joins are deferred to the end
 *   of the block so the user sees content shrink-to-empty in place
 *   first, then lines collapse upward.
 *
 * --mode interleaved:
 *   Delete each line's content, then immediately join the empty line.
 *   delete(line1) -> join -> delete(line2 at L+1) -> join -> ...
 *   Advantage: more incremental — each line disappears completely
 *   before the next is touched.
 *
 * Sliding-window algorithm (Phase 2 fix, handles arbitrary N lines):
 *   Walk the ops. When a `join_lines(L)` is followed by content
 *   deletes at col 1, start a block. Within the block, a 2-line
 *   window [join + delete@col1+] is matched repeatedly. Each match:
 *     - Emits the content deletes at L + 1 + offset (offset = number
 *       of content batches already emitted in this block).
 *     - Defers the join to a pending list.
 *     - Advances offset by 1.
 *   When the block ends (no more join+delete pattern, or a trailing
 *   join with no content), all pending joins are emitted at the end.
 *
 * Only matches when content deletes are at col 1 (full line deletion).
 * Partial content at col > 1 is left unchanged.
 *
 * Build: make layers
 * Usage:  ad_layer_line_delete_in_place [--mode batch|interleaved] < ops.tsv
 */
#include "ad_layer_common.h"

static int ldi_mode = 0;  /* 0=batch, 1=interleaved */

static int layer_line_delete_in_place(Op *ops, int n_ops, Op *out, int out_cap, int *line_offset) {
    (void)line_offset;
    int n_out = 0;
    int i = 0;

    while (i < n_ops) {
        /* Check: is work[i] a join_lines followed by content deletes?
         * This is the start of a potential multi-line block. */
        if (strcmp(ops[i].type, "join_lines") == 0
            && i + 1 < n_ops
            && strcmp(ops[i + 1].type, "delete") == 0
            && !ad_layer_is_line_op(&ops[i + 1])
            && ops[i + 1].col == 1) {

            /* ── Sliding-window block ──
             * Collect content deletes and joins. Emit content first,
             * defer joins to the end of the block. */
            int line_off = 0;  /* number of content batches emitted so far */

            /* Collect joins in a local array (max reasonable block size). */
            Op pending_joins[4096];
            int n_pending = 0;

            int j = i;
            while (j < n_ops && strcmp(ops[j].type, "join_lines") == 0) {
                int join_line = ops[j].line;

                    /* Scan content deletes after this join. The FIRST delete
                 * must be at col 1 (full-line deletion, not partial).
                 * Subsequent deletes can be at any col (they're the rest
                 * of the line's content). */
                int ce = j + 1;
                if (ce < n_ops
                    && strcmp(ops[ce].type, "delete") == 0
                    && !ad_layer_is_line_op(&ops[ce])
                    && ops[ce].col == 1) {
                    ce++;
                    while (ce < n_ops
                           && strcmp(ops[ce].type, "delete") == 0
                           && !ad_layer_is_line_op(&ops[ce]))
                        ce++;
                }

                if (ce > j + 1) {
                    /* Content deletes found (first at col 1).
                     * Emit at join_line + 1 + line_off. */
                    for (int k = j + 1; k < ce && n_out < out_cap; k++) {
                        Op tmp = ops[k];
                        tmp.line = join_line + 1 + line_off;
                        out[n_out++] = tmp;
                    }
                    /* Save the join for deferred emission. */
                    if (n_pending < 4096)
                        pending_joins[n_pending++] = ops[j];
                    line_off++;
                    j = ce;  /* advance past content deletes */
                } else {
                    /* No content deletes after this join — trailing join.
                     * Save it and end the block. */
                    if (n_pending < 4096)
                        pending_joins[n_pending++] = ops[j];
                    j++;
                    break;  /* end of block */
                }
            }

            /* Emit all pending joins (at their original lines). */
            for (int k = 0; k < n_pending && n_out < out_cap; k++)
                out[n_out++] = pending_joins[k];

            i = j;
            continue;
        }

        /* No match — emit unchanged */
        if (n_out < out_cap)
            out[n_out++] = ops[i];
        i++;
    }

    return n_out;
}

/* Interleaved mode pass-through: emit ops unchanged. */
static int layer_passthrough(Op *ops, int n_ops, Op *out, int out_cap, int *line_offset) {
    (void)line_offset;
    int n = (n_ops < out_cap) ? n_ops : out_cap;
    for (int i = 0; i < n; i++)
        out[i] = ops[i];
    return n;
}

int main(int argc, char **argv) {
    __argc = argc; __argv = argv;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--mode") == 0 && i + 1 < argc) {
            i++;
            if (strcmp(argv[i], "interleaved") == 0)
                ldi_mode = 1;
            else
                ldi_mode = 0;  /* batch (default) */
        } else if (strncmp(argv[i], "--mode=", 7) == 0) {
            if (strcmp(argv[i] + 7, "interleaved") == 0)
                ldi_mode = 1;
            else
                ldi_mode = 0;
        } else if (strcmp(argv[i], "--help") == 0 || strcmp(argv[i], "-h") == 0) {
            fprintf(stderr,
                "ad_layer_line_delete_in_place — delete content before joining lines\n\n"
                "Usage: ad_layer_line_delete_in_place [--mode batch|interleaved] < ops.tsv\n\n"
                "Options:\n"
                "  --mode batch        Delete all content first, then join all (default)\n"
                "  --mode interleaved  Delete each line then immediately join\n"
                "  --help, -h          Show this help\n\n"
                "Pattern: join_lines(L) + delete(content at col 1) + join_lines(L) + ...\n"
                "Handles arbitrary N-line deletions via a sliding 2-line window.\n"
                "Only matches full-line deletions (col 1). Partial content at col > 1\n"
                "is left unchanged.\n");
            return 0;
        }
    }

    if (ldi_mode == 1)
        return ad_layer_run(layer_passthrough);

    return ad_layer_run(layer_line_delete_in_place);
}
